Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -1404,6 +1404,127 @@ def visit_ExprNode(self, node):
         return node
 
 
+class _GeneratorExpressionArgumentsMarker(TreeVisitor, SkipDeclarations):
+    def __init__(self, gen_expr):
+        super(_GeneratorExpressionArgumentsMarker, self).__init__()
+        self.gen_expr = gen_expr
+        self.tag = 0
+
+    def visit_ExprNode(self, node):
+        if not node.is_literal:
+            # Don't bother tagging literal nodes
+            assert (not node.generator_arg_tag)  # nobody has tagged this first
+            node.generator_arg_tag = (self.tag, self.gen_expr)
+            self.tag +=1
+        self.visitchildren(node)
+
+    def visit_Node(self, node):
+        return  # we're only interested in the expressions that make up the iterator sequence
+                # so don't go beyond ExprNodes (e.g. into ForFromStatNode)
+
+    def visit_GeneratorExpressionNode(self, node):
+        node.generator_arg_tag = (self.tag, self.gen_expr)
+        self.tag += 1
+        # don't visit children, can't handle overlapping tags
+        # (and assume generator expressions don't end up optimized out in a way
+        #  that would require overlapping tags)
+
+
+class MarkGeneratorExpressionArguments(VisitorTransform, SkipDeclarations):
+    def visit_GeneratorExpressionNode(self, node):
+        self.visitchildren(node)
+        if not isinstance(node.loop, Nodes.ForInStatNode):
+            # Possibly should handle ForFromStatNode
+            # but for now do nothing
+            return node
+        itseq = node.loop.iterator.sequence
+        # literals do not need replacing with an argument
+        if itseq.is_literal:
+            return node
+        _GeneratorExpressionArgumentsMarker(node).visit(itseq)
+        return node
+
+    visit_Node = VisitorTransform.recurse_to_children
+
+
+class HandleGeneratorArguments(VisitorTransform, SkipDeclarations):
```

Comment:
Hmm. Does this need a separate transform? Isn't the outermost iterable special enough to find it without deep traversal?

Reply:
I've moved this into `CreateClosureClasses` so it doesn't have to run separately (which is think is what you were after?)

I've done the same thing for the transform that tags the iterable initially - it's now part of `MarkClosureVisitor`